File: graph_algorithms/week3_shortest_paths/2_bipartite/bipartite.cpp
#include <iostream>
#include <queue>
#include <vector>

using std::queue;
using std::vector;

int bipartite(vector<vector<int> > &adj) {
  vector<int> colour(adj.size(), -1);
  colour[0] = 0;
  queue<int> q;
  q.push(0);
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    for (int v : adj[u]) {
      if (colour[v] == -1) {
        colour[v] = 1 - colour[u];
        q.push(v);
      } else if (colour[v] == colour[u])
        return 0;
    }
  }
  return 1;
}

int main() {
  int n, m;
  std::cin >> n >> m;

  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }

  std::cout << bipartite(adj) << std::endl;
}

File: graph_algorithms/week3_shortest_paths/2_bipartite/bipartite.py
from queue import Queue
from typing import List


def bipartite(adj: List[List[int]]) -> int:  # noqa: C901
    colour = [-1] * len(adj)
    colour[0] = 0
    queue: Queue = Queue()
    queue.put(0)
    while not queue.empty():
        u = queue.get()
        for v in adj[u]:
            if colour[v] == -1:
                colour[v] = 1 - colour[u]
                queue.put(v)
            elif colour[v] == colour[u]:
                return 0
    return 1


if __name__ == "__main__":
    n, m = list(map(int, input().split()))

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)

    print(bipartite(adj))
